#!/usr/bin/python3

import subprocess
import argparse

import os, sys
sys.path.append(os.path.expanduser('~/scripts/python/pylib'))

from myutils import FileSize

def get_output_as_lines(cmd):
  output = subprocess.check_output(cmd)
  return output.decode('utf-8', errors='replace').splitlines()

class QuotaRecord:
  qgid = 'qgroupid'
  rfer = 'rfer'
  excl = 'excl'
  path = 'path'

  def __init__(self, line):
    self.qgid, self.rfer, self.excl, self.path = line.split()
    self.rfer = FileSize(self.rfer)
    self.excl = FileSize(self.excl)

def main(path, sortby):
  output = get_output_as_lines(["btrfs", "qgroup", "show", "--raw", path])

  fmt = '{0.qgid:9}{0.rfer!s:10}{0.excl!s:10}   {0.path}'
  data = [QuotaRecord(l) for l in output[2:]]

  if sortby:
    data = sorted(data, key=lambda d: getattr(d, sortby))
  print(fmt.format(QuotaRecord))
  for d in data:
    print(fmt.format(d))

if __name__ == '__main__':
  parser = argparse.ArgumentParser(
    description='Show BTRFS quota groups in human-readable format')
  parser.add_argument('path', metavar='<path>',
                      help='path of the BTRFS filesystem')
  parser.add_argument('-s', '--sort', choices=('rfer', 'excl', 'path'),
                      help='field to be sorted by')
  args = parser.parse_args()
  try:
    main(args.path, args.sort)
  except subprocess.CalledProcessError as e:
    sys.exit(e.returncode)
